streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala

```diff
@@ -308,12 +305,41 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
         supervisor.start()
         supervisor.awaitTermination()
       }
+
       // Run the dummy Spark job to ensure that all slaves have registered.
       // This avoids all the receivers to be scheduled on the same node.
       if (!ssc.sparkContext.isLocal) {
         ssc.sparkContext.makeRDD(1 to 50, 50).map(x => (x, 1)).reduceByKey(_ + _, 20).collect()
       }
 
+      // Right now, we only honor preferences if all receivers have them
+      val hasLocationPreferences = receivers.map(_.preferredLocation.isDefined).reduce(_ && _)
+
+      // If no location preferences are specified, set host location for each receiver
+      // so as to distribute them evenly over executors in a round-robin fashion
+      var roundRobin = false;
+      if (!hasLocationPreferences && !ssc.sparkContext.isLocal) {
+        val executors = getExecutors(ssc)
+        if (!executors.isEmpty) {
+          var i = 0;
+          for (i <- 0 to (receivers.length - 1)) {
+            receivers(i).host = Some(executors(i % executors.length))
+          }
+          roundRobin = true
+        }
+      }
+
+      val tempRDD =
+        if (hasLocationPreferences) {
```

Comment:
Well, we want to make as little assumption as possible. Users may want to mix Flume receivers (which may have preferred locations) with Kafka Receivers (without any preferred location). So when we are fixing things, its best to fix it in the right way.